eventuate-cdc-data-producer-wrappers/src/main/java/io/eventuate/cdc/producer/wrappers/kafka/TopicPartitionSender.java

```diff
@@ -0,0 +1,161 @@
+        sendMessage();
+      }
+    }
+
+    return topicPartitionMessage.getFuture();
+  }
+
+  private void sendMessage() {
+    if (!messages.isEmpty()) {
+      sendNextMessage();
+    }
+    else {
+      state.set(TopicPartitionSenderState.IDLE);
+
+      /*
+        Additional check is necessary because,
+
+        there can be the case when we got null message and not yet turned state to IDLE.
+        But other thread added message to queue (just after we got null message) and tried to turn state to SENDING.
+        Because state is not yet IDLE, 'sendMessage' will not be invoked.
+        Then state will be turned to IDLE.
+        So message will stay in queue until next 'sendMessage' invocation.
+      */
+
+      if (!messages.isEmpty() && state.compareAndSet(TopicPartitionSenderState.IDLE, TopicPartitionSenderState.SENDING)) {
+        sendMessage();
+      }
+    }
+  }
+
+  private void sendNextMessage() {
+    if (enableBatchProcessing) {
+      sendMessageBatch();
+    }
+    else {
+      sendSingleMessage();
+    }
+  }
+
+  private void sendSingleMessage() {
+    TopicPartitionMessage topicPartitionMessage;
+
+    while ((topicPartitionMessage = messages.poll()) != null) {
+      TopicPartitionMessage message = topicPartitionMessage;
+
+      eventuateKafkaProducer
+        .send(message.getTopic(), message.getKey(), StringUtils.stringToBytes(message.getBody()))
+        .whenComplete((o, throwable) -> {
+          updateMetrics(1);
+          if (throwable != null) {
+            state.set(TopicPartitionSenderState.ERROR);
+            message.completeExceptionally(throwable);
+          }
+          else {
+            message.complete(o);
+            sendMessage();
+          }
+        });
+    }
+  }
+
+  private void sendMessageBatch() {
+    List<TopicPartitionMessage> batch = new ArrayList<>();
+    EventuateKafkaMultiMessageConverter.MessageBuilder messageBuilder = new EventuateKafkaMultiMessageConverter.MessageBuilder(batchSize);
+
+    String key = null;
+    String topic = null;
+
+    AtomicInteger messageCounter = new AtomicInteger(0);
+
+    while (true) {
+      TopicPartitionMessage messageForBatch = messages.peek();
+
+      if (messageForBatch != null &&
+              messageBuilder.addMessage(new EventuateKafkaMultiMessageKeyValue(messageForBatch.getKey(), messageForBatch.getBody()))) {
+
+        messageForBatch = messages.poll();
+
+        //key of the first message is a kafka record key
+        if (key == null) {
+          key = messageForBatch.getKey();
+          topic = messageForBatch.getTopic();
+        }
+
+        batch.add(messageForBatch);
+        messageCounter.incrementAndGet();
+      }
+      else {
+        break;
+      }
+    }
+
+    if (batch.isEmpty()) {
+      state.set(TopicPartitionSenderState.ERROR);
+      throw new RuntimeException("Sender is in error state, publishing is not possible.");
```

Comment:
This error message is not correct- the problem is that the message didn't fit

Reply:
Sorry, I will fix, I just got it from your example: https://github.com/eventuate-foundation/eventuate-cdc/pull/30#discussion_r354931748